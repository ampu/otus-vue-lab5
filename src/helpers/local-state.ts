import {ref, watch} from 'vue'

export type Preprocess<TState, TInput = any> = (state: TInput) => TState;

export class LocalState {
  constructor(
    private key: string,
    private storage: Storage = localStorage,
  ) {
  }

  getState<TState>(defaultState: TState | undefined = undefined) {
    try {
      const item = this.storage.getItem(this.key)
      return item === null
        ? defaultState
        : JSON.parse(item)
    } catch {
      return defaultState
    }
  }

  setState(newState: unknown) {
    try {
      this.storage.setItem(this.key, JSON.stringify(newState))
      return true
    } catch {
      return false
    }
  }

  useState<TState extends object>(preprocess: Preprocess<TState>) {
    const state = ref(preprocess(this.getState()))
    watch(state, () => this.setState(state.value), {deep: true})
    return state
  }
}

// new LocalState(`otus-vue.books`).useState<BookModel[]>((state) => state ?? [])
