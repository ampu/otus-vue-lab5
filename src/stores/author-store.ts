import {ref} from 'vue'
import {defineStore} from 'pinia'

import type {AuthorFilterModel, AuthorModel} from '@/utils/author-types'

export const useAuthorStore = defineStore(`author`, () => {
  const authors = ref([] as AuthorModel[])

  const getAuthors = (filter?: AuthorFilterModel) => {
    if (!filter) {
      return authors.value
    }
    const {name, country} = filter
    const lcName = name?.toLowerCase()
    const lcCountry = country?.toLowerCase()
    return authors.value.filter((author) => (
      (!lcName || author.name.toLowerCase().includes(lcName))
      && (!lcCountry || author.country?.toLowerCase().includes(lcCountry))
    ))
  }

  const getAuthor = (authorId: string) => authors.value.find((author) => author.id === authorId)

  const addAuthor = async (author: AuthorModel) => {
    authors.value.push(author)
  }

  const updateAuthor = async (author: AuthorModel) => {
    const index = authors.value.findIndex(({id}) => id === author.id)
    if (index !== -1) {
      authors.value[index] = author
    }
  }

  return {
    authors,
    getAuthors,
    getAuthor,
    addAuthor,
    updateAuthor,
  }
})
