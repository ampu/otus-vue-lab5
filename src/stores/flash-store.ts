import {ref} from 'vue'
import {defineStore} from 'pinia'
import {nanoid} from 'nanoid'

import type {FlashModel, FlashStatus} from '@/utils/flash-types'

const DEFAULT_TIMEOUT = 3000
const SCHEDULER_INTERVAL = 1000

export const useFlashStore = defineStore(`flash`, () => {
  const flashes = ref([] as FlashModel[])

  const addFlash = ({
    status,
    message,
    ttl = DEFAULT_TIMEOUT,
  }: {
    status: FlashStatus,
    message: string
    ttl?: number
  }) => {
    flashes.value.push({
      id: nanoid(),
      status,
      message,
      deadTimestamp: performance.now() + ttl
    })
  }

  const removeFlash = ({id}: FlashModel) => {
    flashes.value = flashes.value.filter((flash) => flash.id !== id)
  }

  const runScheduler = () => {
    setInterval(() => {
      const now = performance.now()
      const predicate = (flash: FlashModel) => now < flash.deadTimestamp
      if (!flashes.value.every(predicate)) {
        flashes.value = flashes.value.filter(predicate)
      }
    }, SCHEDULER_INTERVAL)
  }

  runScheduler()

  return {
    flashes,
    addFlash,
    removeFlash,
  }
})
