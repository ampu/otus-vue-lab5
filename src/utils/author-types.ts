export interface AuthorModel {
  id: string
  name: string
  birthday?: string
  sex?: `male` | `female`
  country?: string
}

export interface AuthorFilterModel {
  name?: string
  country?: string
}
