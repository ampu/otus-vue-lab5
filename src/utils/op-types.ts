export const enum OpStatus {
  NONE = 'NONE',
  SUCCESS = 'SUCCESS',
  ERROR = 'ERROR',
  PENDING = 'PENDING',
}

export interface OpStatuses {
  disabled: boolean
  active: boolean
  pending: boolean
  error: boolean
}

export const getStatuses = (status: OpStatus): OpStatuses => {
  return {
    disabled: status !== OpStatus.SUCCESS,
    active: status === OpStatus.SUCCESS,
    pending: status === OpStatus.PENDING,
    error: status === OpStatus.ERROR,
  }
}
