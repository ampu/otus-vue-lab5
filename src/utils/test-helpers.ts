import {expect, vi} from 'vitest'
import {DOMWrapper} from '@vue/test-utils'
import {RouteLocationRaw} from 'vue-router'

import router from '@/router'

export const stubScrollTo = () => {
  window.scrollTo = vi.fn() as any
}

export const expectToHaveDisabledAttribute = <TNode extends Node>(wrapper: DOMWrapper<TNode>) => {
  expect(wrapper.attributes(`disabled`)).toBeDefined()
}

export const clickAndExpectRouterPush = async (wrapper: DOMWrapper<Element>, route: RouteLocationRaw, shouldWaitForNavigation = false) => {
  const pushSpy = vi.spyOn(router, `push`)
  await wrapper.trigger(`click`)
  expect(pushSpy).toHaveBeenCalledWith(expect.objectContaining(route))
  pushSpy.mockClear()

  if (shouldWaitForNavigation) {
    await router.push(route)
    expect(router.currentRoute.value).toEqual(expect.objectContaining(route))
  }
}
