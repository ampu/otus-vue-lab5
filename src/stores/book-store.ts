import {computed, ref} from 'vue'
import {defineStore} from 'pinia'

import type {BookFilterModel, BookModel} from '@/utils/book-types'
import bookClient from '@/clients/book-client'
import {useAuthorStore} from '@/stores/author-store'
import {getStatuses, OpStatus} from '@/utils/op-types'

export const useBookStore = defineStore(`book`, () => {
  const fetchStatus = ref(OpStatus.NONE)
  const books = ref([] as BookModel[])

  const fetchStatuses = computed(() => getStatuses(fetchStatus.value))

  const getBook = (bookId: string) => {
    return books.value.find((book) => book.id === bookId)
  }

  const addBook = async (book: BookModel) => {
    books.value.push(book)
  }

  const updateBook = async (book: BookModel) => {
    const index = books.value.findIndex(({id}) => id === book.id)
    if (index !== -1) {
      books.value[index] = book
    }
  }

  const removeBook = async ({id}: BookModel) => {
    books.value = books.value.filter((book) => book.id !== id)
  }

  const getBooks = (filter?: BookFilterModel) => {
    if (!filter) {
      return books.value
    }
    const {
      title,
      minYear,
      maxYear,
      category,
      minPrice,
      maxPrice,
    } = filter
    const lcTitle = title?.toLowerCase()
    const lcCategory = category?.toLowerCase()
    return books.value.filter((book) =>
      (!lcTitle || book.title?.toLowerCase().includes(lcTitle))
      && (!minYear || minYear <= book.year!)
      && (!maxYear || book.year! <= maxYear)
      && (!lcCategory || book.category?.toLowerCase().includes(lcCategory))
      && (!minPrice || minPrice <= book.price!)
      && (!maxPrice || book.price! <= maxPrice))
  }

  const fetchBooks = async () => {
    if ([OpStatus.PENDING, OpStatus.SUCCESS].includes(fetchStatus.value)) {
      return
    }
    try {
      fetchStatus.value = OpStatus.PENDING
      const {books, authors} = await bookClient.getBooks(`vue.js`)
      books.forEach(addBook)

      const authorStore = useAuthorStore()
      authors.forEach(authorStore.addAuthor)
      fetchStatus.value = OpStatus.SUCCESS
    } catch (err) {
      fetchStatus.value = OpStatus.ERROR
      console.error((err as Error).stack)
    }
  }

  return {
    fetchStatus,
    fetchStatuses,
    books,
    getBook,
    addBook,
    updateBook,
    removeBook,
    getBooks,
    fetchBooks,
  }
})
