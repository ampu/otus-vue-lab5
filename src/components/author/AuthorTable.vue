<template>
  <table class="table table-striped align-middle text-center">
    <thead>
      <AuthorTableHeader/>
      <AuthorFilter
        :disabled="statuses.disabled"
        v-model="filter"
      />
    </thead>

    <TableStatuses
      title="authors"
      :statuses="statuses"
      :isEmpty="isEmpty"
      :isEmptyByFilter="isEmptyByFilter"
    />

    <Transition name="fade" mode="out-in">
      <tbody
        v-if="statuses.active"
        :key="bodyKey"
      >
        <AuthorTableRow
          v-for="author of pageAuthors"
          :key="author.id"
          :author="author"
        />
      </tbody>
    </Transition>

    <TableFooter
      v-if="statuses.active"
      :isFiltered="isFiltered"
      :total="authorStore.authors.length"
      :filtered="filteredAuthors.length"
      :perPage="perPage"
      v-model:page="page"
    />
  </table>
</template>

<script setup lang="ts">
import {onBeforeMount, computed, ref} from 'vue'

import TableFooter from '@/components/TableFooter.vue'
import AuthorTableRow from '@/components/author/AuthorTableRow.vue'

import {useAuthorStore} from '@/stores/author-store'
import {useBookStore} from '@/stores/book-store'
import {AuthorFilterModel} from '@/helpers/author-types'
import {getLatestPage, getPageItems} from '@/helpers/navigation-helpers'
import {usePage} from '@/hooks/use-page'
import TableStatuses from '@/components/TableStatuses.vue'
import AuthorTableHeader from '@/components/author/AuthorTableHeader.vue'
import AuthorFilter from '@/components/author/AuthorFilter.vue'

const authorStore = useAuthorStore()
const bookStore = useBookStore()

const statuses = computed(() => bookStore.fetchStatuses)
const filter = ref({} as AuthorFilterModel)
const isFiltered = computed(() => Object.values(filter.value).some(Boolean))
const filteredAuthors = computed(() => authorStore.getAuthors(filter.value))
const isEmpty = computed(() => authorStore.authors.length === 0)
const isEmptyByFilter = computed(() => filteredAuthors.value.length === 0)
const perPage = ref(5)
const page = usePage(`authors`, () => getLatestPage(filteredAuthors.value.length, perPage.value))
const pageAuthors = computed(() => getPageItems(filteredAuthors.value, page.value, perPage.value))
const bodyKey = computed(() => page.value)

onBeforeMount(bookStore.fetchBooks)
</script>

<style lang="scss" scoped>
</style>
