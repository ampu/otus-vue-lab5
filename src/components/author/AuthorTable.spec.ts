import {describe, it, expect, vi, afterEach} from 'vitest'
import {mount} from '@vue/test-utils'

import AuthorTable from '@/components/author/AuthorTable.vue'
import {AuthorModel} from '@/helpers/author-types'
import {createTestingPinia} from '@pinia/testing'
import router from '@/router'
import {useBookStore} from '@/stores/book-store'
import {OpStatus} from '@/helpers/op-types'
import {clickAndExpectRouterPush} from '@/helpers/test-helpers'

describe(`AuthorTable`, () => {
  const ITEMS: AuthorModel[] = [
    {id: `1`, name: `Author #1`},
    {id: `2`, name: `Author #2`, sex: `male`},
    {id: `3`, name: `Author #3`, country: `Germany`},
    {id: `4`, name: `Author #4`, birthday: `1987-08-26`},
    {id: `5`, name: `Author #5`, sex: `female`, country: `Poland`},
    {id: `6`, name: `Author #6`, sex: `male`, birthday: `1900-01-01`},
    {id: `7`, name: `Author #7`, sex: `female`, country: `Israel`, birthday: `1982-08-19`},
  ]

  const mountComponent = (
    fetchStatus = OpStatus.NONE,
    authors = [] as AuthorModel[],
  ) => {
    const pinia = createTestingPinia({
      stubActions: false,
      initialState: {
        book: {
          fetchStatus,
        },
        author: {
          authors,
        }
      },
    })

    const bookStore = useBookStore()
    bookStore.fetchBooks = vi.fn()

    return mount(AuthorTable, {
      global: {
        plugins: [router, pinia],
      }
    })
  }

  afterEach(() => {
    vi.restoreAllMocks()
  })

  it(`imports without error`, () => {
    expect(AuthorTable).toBeTypeOf(`object`)
  })

  it(`mounts without errors`, () => {
    const wrapper = mountComponent(undefined, ITEMS)
    expect(wrapper.exists()).toBe(true)
  })

  it(`renders by default`, () => {
    const wrapper = mountComponent(undefined, ITEMS)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it(`renders list`, () => {
    const wrapper = mountComponent(OpStatus.SUCCESS, ITEMS)

    const itemWrappers = wrapper.findAll(`[data-testid^="author-item-"]`)
    expect(itemWrappers.map((item) => item.text())).toMatchSnapshot()
  })

  it(`navigates to author edit`, async () => {
    const wrapper = mountComponent(OpStatus.SUCCESS, ITEMS)
    const editWrapper = wrapper.find(`[data-testid="author-edit-1"]`)
    expect(editWrapper.exists()).toBe(true)

    await clickAndExpectRouterPush(editWrapper, {name: `authors-edit`, params: {authorId: '1'}})
  })

  it(`navigates to pages`, async () => {
    const wrapper = mountComponent(OpStatus.SUCCESS, ITEMS)
    const pageWrappers = wrapper.findAll(`[data-testid="pagination-page"]`)
    expect(pageWrappers).toHaveLength(4)

    const lastPageWrapper = pageWrappers[2]
    expect(lastPageWrapper.text()).toBe(`2`)

    const route = {
      name: `authors`,
      query: {page: `2`},
    }
    await clickAndExpectRouterPush(lastPageWrapper, route, true)

    const itemWrappers = wrapper.findAll(`[data-testid^="author-item-"]`)
    expect(itemWrappers.map((item) => item.text())).toMatchSnapshot()
  })
})
