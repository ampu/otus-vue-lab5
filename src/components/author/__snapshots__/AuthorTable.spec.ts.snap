// Vitest Snapshot v1

exports[`AuthorTable > navigates to pages 1`] = `
[
  "Author #6-",
  "Author #7Israel",
]
`;

exports[`AuthorTable > renders by default 1`] = `
"<table class=\\"table table-striped align-middle text-center\\">
  <thead>
    <tr class=\\"title\\">
      <th>Name</th>
      <th>Country</th>
      <th style=\\"width: 58px;\\"></th>
    </tr>
    <tr>
      <th><input class=\\"form-control\\" disabled=\\"\\"></th>
      <th><input class=\\"form-control\\" disabled=\\"\\"></th>
      <th></th>
    </tr>
  </thead>
  <!--v-if-->
  <!--v-if-->
  <!--v-if-->
  <transition-stub name=\\"fade\\" mode=\\"out-in\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
    <!--v-if-->
  </transition-stub>
  <!--v-if-->
</table>"
`;

exports[`AuthorTable > renders list 1`] = `
[
  "Author #1-",
  "Author #2-",
  "Author #3Germany",
  "Author #4-",
  "Author #5Poland",
]
`;
